Issue 9854 implement site creation dynamic texts experience

steineron opened the pull request:
closes #9854 :
Display a more dynamic preview for site creation experience, by sequencing several "status" texts while the site is being created. Currently one static text is displayed and this PR is about making this text change dynamically while the user waits on the server's response.

The texts originate from a constant list/series and are displayed in a cyclic manner as time passes.

The proposal for this work is here: p77Llu-d6r-p2


Texts need translations to other resourced languages. 

Test suggestions:

Either it's hard or impossible to test this with development builds - for some reason I got 403 error on every attempt, possibly to avoid creation on infinite pointless sites by devs/testers(?). So I used Charles proxy to control the response:
```
{"success":true,"blog_details":{"url":"https:\/\/trialsite.wordpress.com\/","blogid":"170723916","blogname":"Ron Trial Site","xmlrpc":"https:\/\/trialsite.wordpress.com\/xmlrpc.php"}}
```

Create a site and figure a way to stall the api response (for that new site creation request), e.g. use Charles proxy and breakpoint requests to `public-api.wordpress.com/rest/v1.1/sites/new/*`. While the request is waiting for the response - watch the texts sequence play.

(If using proxy, consider mocking different outcomes to the request - fail and success to make sure the sequence terminates properly)

Also, on fail, use the `Retry` feature to verify that the sequence is (re)played properly.

See the following vids that highlight the gist of it (ps: for some reason the gifs are slooooowwwwww, where IRL it's 2 seconds per text):

![9854-3_1](https://user-images.githubusercontent.com/3023254/73538080-c7131380-4475-11ea-9eed-6de5d9f859b4.gif)
![9854-4](https://user-images.githubusercontent.com/3023254/73538088-ca0e0400-4475-11ea-9185-32167c2a4289.gif)


PR submission checklist:

- [v] I have considered adding unit tests where possible.
- [v] I have considered adding accessibility improvements for my changes.
- [v] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

steineron commented on WordPress/src/main/java/org/wordpress/android/ui/sitecreation/previews/SitePreviewViewModel.kt:
```diff
@@ -318,7 +318,13 @@ class SitePreviewViewModel @Inject constructor(
                 fullscreenProgressLayoutVisibility = true
         ) {
-            const val loadingTextResId = R.string.notification_new_site_creation_creating_site_subtitle
+            val loadingTextResId:List<Int> = listOf(
```
convert the previously single-text to an array of texts (res-ids)

steineron commented on WordPress/src/main/java/org/wordpress/android/ui/sitecreation/previews/SitePreviewViewModel.kt:
```diff
@@ -318,7 +318,13 @@ class SitePreviewViewModel @Inject constructor(
                 fullscreenProgressLayoutVisibility = true
         ) {
-            const val loadingTextResId = R.string.notification_new_site_creation_creating_site_subtitle
+            val loadingTextResId: List<Int> = listOf(
```
Replace the single-valued res-id on hte view-model with an ordered collection of the texts to be displayed

steineron commented on WordPress/src/main/java/org/wordpress/android/ui/sitecreation/previews/SiteCreationPreviewFragment.kt:
```diff
@@ -214,14 +246,27 @@ class SiteCreationPreviewFragment : SiteCreationBaseFormFragment(),
+            fullscreenProgressLayout.findViewById<TextView>(R.id.progress_text)?.let { textView ->
+
+                // create a progress helper and let it run
+                SiteCreationTextsProgressHelper(WeakReference(textView), loadingTextResId).also {
```
👍

steineron commented:
@malinajirka  - thanks for the welcomed feeddback. To your comments:
1. `closes` is an issue-closing keyword just as `fixes` is AFAIK ;)
2. I see your point about the fade in/out animations and I'll added the fade - I don't know how I've missed that one.

steineron commented:
@malinajirka  - After some poking and thinking - I've ended up with the following solution:
1. A `TextSwitcher` fades the texts in/out. 
2. a call to `postDelayed` times the next call to change a text

I'm looking into unit-testing the logic in `SiteCreationTextsProgressHelper`

steineron commented on WordPress/src/main/java/org/wordpress/android/ui/sitecreation/previews/SiteCreationPreviewFragment.kt:
```diff
@@ -127,17 +142,39 @@ class SiteCreationPreviewFragment : SiteCreationBaseFormFragment(),
                 uiHelpers.updateVisibility(sitePreviewWebError, uiState.webViewErrorVisibility)
-                uiHelpers.updateVisibility(sitePreviewWebViewShimmerLayout, uiState.shimmerVisibility)
-                uiHelpers.updateVisibility(fullscreenErrorLayout, uiState.fullscreenErrorLayoutVisibility)
+                uiHelpers.updateVisibility(
```
...hmm...
did not intend to reformat this file(s). @malinajirka  - if that's unacceptable - I'll revert that change (format)

steineron replied:
(I'm fixing it - it'll be in the next commit)

steineron commented on WordPress/src/main/java/org/wordpress/android/ui/sitecreation/previews/SiteCreationPreviewFragment.kt:
```diff
@@ -356,6 +413,91 @@ class SiteCreationPreviewFragment : SiteCreationBaseFormFragment(),
+                            .inflate(R.layout.site_creation_progress_text, this, false)
+                }
+                // set fade in/out animations
+                inAnimation = AnimationUtils.loadAnimation(context, R.anim.fade_in)
```
using existing fade-in/out animations (i.e. already defined in the project) @shaktimb - could you advise whether these animations are fine? they run for 600ms.

steineron commented on WordPress/src/main/java/org/wordpress/android/ui/sitecreation/previews/SiteCreationPreviewFragment.kt:
```diff
@@ -356,6 +413,91 @@ class SiteCreationPreviewFragment : SiteCreationBaseFormFragment(),
+            if (canceled) {
+                return
+            }
+            textSwitcher.get()?.let {
```
that's a mistake - I'll fix it with the next commit (no need to get it twice)

steineron commented on WordPress/src/main/res/layout/progress_layout.xml:
```diff
@@ -9,13 +9,11 @@
+    <TextSwitcher
         android:id="@+id/progress_text"
-        android:layout_width="wrap_content"
+        android:layout_width="match_parent"
```
`match_parent` improves flickering of layout since different texts have different lengths and it creates an odd effect with `TextSwitcher` (it uses 2 `TextView` instance under the hood)

steineron commented on WordPress/src/androidTest/java/org/wordpress/android/util/helpers/DynamicTextsProgressionHelperTest.kt:
```diff
@@ -0,0 +1,250 @@
+            }
+        }
+        // verify
+        onView(withText(R.string.notification_new_site_creation_creating_site_dynamic_subtitle_1))
```
because there's only a `TextSwitcher` - every second the text on the screen should change - or the test will fail (`onView(withText(...))` will fail to match a view)

malinajirka commented on WordPress/src/main/java/org/wordpress/android/ui/sitecreation/previews/SiteCreationPreviewFragment.kt:
```diff
@@ -72,6 +77,9 @@ class SiteCreationPreviewFragment : SiteCreationBaseFormFragment(),
     private var okButtonContainer: View? = null
 
+    // an instance helping sequence texts while in `loading` state
+    private var mTextsProgressTextsHelper: DynamicTextsProgressionHelper? = null
```
We use the hungarian notation only in Java files. It'd be good to remove it for consistency. Thanks!

malinajirka commented on WordPress/src/main/java/org/wordpress/android/ui/sitecreation/previews/SitePreviewViewModel.kt:
```diff
@@ -318,7 +318,13 @@ class SitePreviewViewModel @Inject constructor(
         ) {
-            const val loadingTextResId = R.string.notification_new_site_creation_creating_site_subtitle
+            val loadingTextResIds: List<Int> = listOf(
+                    R.string.notification_new_site_creation_creating_site_dynamic_subtitle_1,
```
We might consider using `UIString` wrapper for consistency with other VMs. Wdyt?

steineron replied:
More than happy to follow existing conventions :)

malinajirka commented on WordPress/src/main/java/org/wordpress/android/util/helpers/DynamicTextsProgressionHelper.kt:
```diff
@@ -0,0 +1,100 @@
+    private val stringIds: List<Int>,
+    val delay: Long
+) : Runnable {
+    val nText: Int = stringIds.size
```
Out of curiosity, is using `n` prefix some kind of a convention? 

Either case, we might want to consider renaming the variable to a more specific name. Wdyt?

steineron replied:
n is mere notation like you'd use in math books... Ntexts.. the number of texts... makes sense? anyway - I'll change it to something clearer :)

malinajirka commented on WordPress/src/main/java/org/wordpress/android/util/helpers/DynamicTextsProgressionHelper.kt:
```diff
@@ -0,0 +1,100 @@
+    val nText: Int = stringIds.size
+
+    // the count of how many times `nextText` was executed
+    var count = -1
```
We might want to consider making all the fields private, wdyt?

steineron replied:
definitely. It's a remainder from when the class was `inner private`

malinajirka commented on WordPress/src/main/java/org/wordpress/android/util/helpers/DynamicTextsProgressionHelper.kt:
```diff
@@ -0,0 +1,100 @@
+
+    init {
+        textSwitcher.get()?.apply {
+            removeAllViews()
```
Are there any views by default? Or why do we need to start by removing all view?

steineron replied:
this class has no knowledge of the `TextSwitcher` instance it receives. if this instance has already been used - it might have child views and the setup following (factory/animations etc) will fail - `TextSwitcher` is built that way. in `SiteCreationPreviewFragment`, failure followed by `Retry` triggers exactly that. That's why it is part of the`TextSwitcher`'s setup

malinajirka commented on WordPress/src/main/java/org/wordpress/android/util/helpers/DynamicTextsProgressionHelper.kt:
```diff
@@ -0,0 +1,100 @@
+            setFactory {
+                // inflate a text view that matches the design for this screen
+                LayoutInflater.from(context)
+                        .inflate(R.layout.site_creation_progress_text, this, false)
```
This resource is site_creation specific. Since all the other things in this class are feature-independent, I'd consider passing it as a parameter. Wdyt?

steineron replied:
spot on with that one - residue from when it was `private inner` in `SiteCreateionP{reviewFramgment`. I'll do it.

steineron commented:
@malinajirka thanks for the thorough review. Appreciated!

>Tbh when we discussed if we should implement this feature in the VM-layer or in the view-layer, I thought the implementation in the view layer would use just the Android SDK


re the animation mechanism, I think both our initial thoughts about harnessing the SDK animations the way you described are a good starting point (I even have a stash with some first steps in that direction). However, once I found that there's a dedicated class exactly for that in the SDK - the `TextSwitcher` - I couldn't resist using it and favour a more _setup_ code over _flow-control_ code. I'm quite happy with this trade, I think it's a good one and I have more confidence in it now that the animations and text changes are done within an SDK unit and my code mostly setup things and triggers texts-change :)

>If we decrease the length of the sleep, the test will be faster, but it'll fail (run indefinitely) more often (it's flaky by its nature). Even though they are well written I'd consider removing them from the above reason and for consistency since we don't have similar test anywhere else

The instrumented unit tests (I save the term UI tests for end-to-end tests that drive the ui the way the users do, but that's just my team's thing) - what I did was to create a unit test, an instrumented unit test that yet uses espresso, that test my class's logic (the unit of logic) in isolation. Now since the animations are disabled for instrumented tests, the _only_ use of `sleep` is to wait until the next pulse/trigger that changes the text, _not_ to sleep while the view animates. That's all. Like I said - `sleep` is merely a tool in your tool-belt and should be used appropriately (def not to sleep through animations - which isn't the case here).
However - I do see you super-valid point that such a small piece of logic can't consume so many seconds in a test that runs with every CI build, so I dropped the duration from 1s/1000ms to 0.1s/100ms (TBH - I feel 0.2s is a better value as such tests are sensitive to the device/emulator's computational strength and so fail at times). 

Since this component _may_ be used elsewhere and somebody _may_ expand/modify it.. there's some value in having it tested and having a test setup so the modifiers can add their own tests as they change the implementation. It is entirely up to you, it's beyond my judgment because I'm a guest in your project. With that said - I would recommend keeping this test *for now*, and would totally get it if you opt not to.

So I've addressed your comments with the latest commit and left the tests till you advise whether you like them there or not :)

Have a great weekend!

steineron commented on WordPress/src/main/java/org/wordpress/android/ui/sitecreation/previews/SiteCreationPreviewFragment.kt:
```diff
@@ -214,7 +232,18 @@ class SiteCreationPreviewFragment : SiteCreationBaseFormFragment(),
     private fun updateLoadingLayout(progressUiState: SitePreviewFullscreenProgressUiState) {
         progressUiState.apply {
-            uiHelpers.setTextOrHide(fullscreenProgressLayout.findViewById(R.id.progress_text), loadingTextResId)
+            fullscreenProgressLayout.findViewById<TextSwitcher>(R.id.progress_text)?.apply {
```
I like this minimal integration with the existing class ;) wdyt @malinajirka?

malinajirka commented:
> However, once I found that there's a dedicated class exactly for that in the SDK - the TextSwitcher - I couldn't resist using it 

I agree that the TextSwitcher is quite a nice component and perfect for the task. 

As I mentioned ` I thought the implementation in the view layer would use just the Android SDK`. When we found out we'd need to introduce a custom logic in the view layer, I think it might have been a good idea to revisit the initial plan and consider passing only a single "resId" from the VM. 

We almost wouldn't need to touch the view layer and we'd just emit textChange every x seconds from the VM - TextSwitcher would take care of the animations. This code could be completely unit tested. Having said, the final solution is also ok. 


>I save the term UI tests for end-to-end tests that drive the ui the way the users do, but that's just my team's thing

Good point. I agree that separating these two cases makes sense.

>what I did was to create a unit test, an instrumented unit test that yet uses espresso, that test my class's logic (the unit of logic) in isolation. 

👍 Yeah, I like the approach how you tested only a single component. 

>However - I do see you super-valid point that such a small piece of logic can't consume so many seconds in a test that runs with every CI build, so I dropped the duration from 1s/1000ms to 0.1s/100ms 

I agree that decreasing the value is definitely a good call. However, the tests still run for 1+s (not including start up time). Which is still quite a lot. But I guess it might be acceptable.

>Since this component may be used elsewhere and somebody may expand/modify it.. there's some value in having it tested and having a test setup so the modifiers can add their own tests as they change the implementation. 

I agree it's nice to have it tested. At the same time I think keeping consistent approaches across the codebase is also really important. Moreover, we both agreed that the test is flaky by its nature which isn't great as the test looses a lot of its value. We can ask on Slack what other devs think about it as this is a very opinionated topic.

jkmassel commented:
We're freezing `14.2` today, so this PR is being bumped to `14.3`. If you'd like it to ship with `14.2`, please merge it into `release/14.2` and ping me – I'll be happy to cut a new beta release.

oguzkocer commented:
We're freezing `14.3` today, so this PR is being bumped to `14.4`. If you'd like it to ship with `14.3`, please merge it into `release/14.3` and ping me – I'll be happy to cut a new beta release.

oguzkocer commented:
We're freezing `14.4` today, so this PR is being bumped to `14.5`. If you'd like it to ship with 14.4, please merge it into release/14.4 and ping me – I'll be happy to cut a new beta release.

oguzkocer commented:
I just realized that this PR is bumped for the 3rd time without any updates. @steineron @malinajirka Could you follow up on it when you get a chance? Thanks!

malinajirka commented:
>I just realized that this PR is bumped for the 3rd time without any updates.

Sorry, I haven't had time to get to it yet. I'd like to do a couple of minor changes before we merge this PR into develop. I've created a new branch and I'm merging these changes there so we don't lose the changes from the forked repo.

Thank you @steineron for working on this feature ;)!
